### ISSUE TITLE
direct access of $_GET/$_POST is not allowed

### ISSUE THREAD

**Issue by danmarsden** [dualcube/moodle-enrol_braintree#1]
please replace these with optional/required_param calls with appropriate foramts such as PARAM_INT, etc.

**Comment by parthajeet** [comment]
Hello Dan,

Thanks for the suggestion. We have replace $_POST with data_submitted.

Regards,
Parthajeet~

**Comment by danmarsden** [comment]
data_submitted() is also raw data - you should not be using that either. Please replace with correct usage of optional/required_param calls with appropriate formats such as PARAM_INT, PARAM_ALPHANUMEXT etc.

**Comment by parthajeet** [comment]
Hello Dan,

I understand your concern. Can you please provide us with an example of how to handle post data from 3rd party service. We have gone through the code of default Moodle paypal plugin and there also its done the same way. As we are not sure of all the response from $_POST so how we can assign variable with optional/required_param for each post values.

Thanks,
Parthajeet

**Comment by danmarsden** [comment]
You must sanitise the data and make sure it is safe. You do not do this when iterating over data_submitted() and storing the data into an array or when saving that data into the database. The core paypal plugin is not an example of good code either and if it was submitted in that state as a new plugin now it's unlikely it would pass through the review process and be integrated in that state.
The variables passed back from the external API are usually documented in the external API docs - many of the variables names and formats should be known as you are storing them directly into the db. I really think you should improve this before we accept this into the plugins db - IMO plugins that have anything to do with payments should be treated as high-risk and therefore should comply well with all security standards and guidelines. However - I do not personally make the call on approval/rejection within the plugins db so if you believe that this code should still be approved in it's current state feel free to post your reasons into a comment on your plugins db entry.

**Comment by mudrd8mz** [comment]
Thanks guys for trying to sort this issue out. I fully agree with Dan here. That "we are not sure of all the response from $_POST" is not something the clients of your plugin would be happy about - especially if their money are on the table. And in fact I believe we can achieve a reliable fix here.

What we would just like to see in the checkout.php is that you parse $_POST while being very strict and picky on what you get. As far as I can see, there are only couple of response values you actually use, and you know their type. So why (instead of blindly iterating over the response) not to pick just those $_POST data you expect and validate them strictly. In fact, you should be even more strict that the default `clean_param()`. If you expect the external system to return integer data, I would even prefer seeing an exception if integer data is not received, rather than silently casting it to int.

In other words, the problem here is not that $_POST is used. And the fix was not supposed to masquerade it via `data_submitted()`. It is perfectly valid to access $_POST in certain cases (such as when data are received back in custom dash-separated format). But you should still be very strict on what input data you use and what type you require.

Once this is implemented properly here, you can apply the same fix in your other payment related plugins - and be proud of offering a code with better-than-standard quality. I believe it is worth of it :-)
